core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1163,6 +1176,11 @@ public void failShard(String reason, @Nullable Throwable e) {
         public ShardRouting routingEntry() {
             return indexShard.routingEntry();
         }
+
+        @Override
+        public long opPrimaryTerm() {
+            return routingEntry().primaryTerm();
+        }
```

Comment:
> A new primary term means that we need to route to a new primary location.

for now primary terms are only used as a way to detect *old* primaries. Later on a new primary term will probably mean we have to resync (though this is not yet fully crystallized). 

> If IndexShard gets created as primary shard (that is not a relocation target), then it should have the same primary term until its closed, right?

Correct. Even if it's a target - for now we don't increment the primary term when we move a primary from one place to another (we have other mechanics to guarantee primary relocation). Only a new primary / primary promotion does it. I agree this may change in the future, but my current thinking is to keep as is it will tigger a resync. 

I will bake these assertions.